File: frontend/src/types/project.ts
import { User } from './auth';

export type ProjectStatus = 'PLANNING' | 'ACTIVE' | 'ON_HOLD' | 'COMPLETED' | 'ARCHIVED';

export interface Project {
  id: string;
  workspace_id: string;
  name: string;
  description?: string | null;
  status: ProjectStatus;
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT';
  deadline?: string | null;
  budget?: number | null;
  owner_id?: string | null;
  created_by_id?: string | null;
  created_at: string;
  updated_at: string;
  owner?: User | null;
  created_by?: User | null;
}

export interface ProjectMember {
  id: string;
  project_id: string;
  user_id: string;
  role: string;
  joined_at: string;
  user?: User;
}

export interface ProjectCreatePayload {
  name: string;
  description?: string;
  status?: ProjectStatus;
  priority?: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT';
  deadline?: string;
  budget?: number;
  owner_id?: string;
}

File: backend/app/schemas/project.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.models.project import ProjectStatus
from app.schemas.user import UserResponse


class ProjectBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=2, max_length=255)
    description: Optional[str] = None
    status: ProjectStatus = ProjectStatus.PLANNING
    priority: str = Field("MEDIUM", pattern="^(LOW|MEDIUM|HIGH|URGENT)$")
    deadline: Optional[datetime] = None
    budget: Optional[float] = Field(None, ge=0)
    owner_id: Optional[str] = None


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = Field(None, min_length=2, max_length=255)
    description: Optional[str] = None
    status: Optional[ProjectStatus] = None
    priority: Optional[str] = Field(None, pattern="^(LOW|MEDIUM|HIGH|URGENT)$")
    deadline: Optional[datetime] = None
    budget: Optional[float] = Field(None, ge=0)
    owner_id: Optional[str] = None


class ProjectResponse(ProjectBase):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    id: str
    workspace_id: str
    created_by_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    owner: Optional[UserResponse] = None
    created_by: Optional[UserResponse] = None


class ProjectMemberCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_id: str
    role: str = Field("MEMBER", min_length=2, max_length=50)


class ProjectMemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    project_id: str
    user_id: str
    role: str
    joined_at: datetime
    user: Optional[UserResponse] = None
